pkg/client/client.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package client
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/projectriff/http-gateway/pkg/liiklus"
+	"github.com/projectriff/http-gateway/pkg/serialization"
+	"google.golang.org/grpc"
+)
+
+// StreamClient allows publishing to a riff stream, through a liiklus gateway and using the riff serialization format.
+type StreamClient struct {
+	// Gateway is the host:port of the liiklus gRPC endpoint.
+	Gateway string
+	// TopicName is the name of the liiklus topic backing the stream.
+	TopicName string
+	// acceptableContentType is the content type that the stream is able to persist. Incompatible content types will be rejected.
+	acceptableContentType string
+	// client is the gRPC client for the liiklus API.
+	client liiklus.LiiklusServiceClient
+	// conn is a reference to the underlying connection, kept for proper cleanup.
+	conn *grpc.ClientConn
+}
+
+type PublishResult struct {
+	Partition uint32
+	Offset    uint64
+}
+
+// NewStreamClient creates a new StreamClient for a given stream.
+func NewStreamClient(gateway string, topic string, acceptableContentType string) (*StreamClient, error) {
+	timeout, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+	conn, err := grpc.DialContext(timeout, gateway, grpc.WithInsecure(), grpc.WithBlock())
+	if err != nil {
+		return nil, err
+	}
+	client := liiklus.NewLiiklusServiceClient(conn)
+	return &StreamClient{
+		Gateway:               gateway,
+		TopicName:             topic,
+		acceptableContentType: acceptableContentType,
+		client:                client,
+		conn:                  conn,
+	}, nil
+}
+
+func (lc *StreamClient) Publish(ctx context.Context, payload io.Reader, key io.Reader, contentType string, headers map[string]string) (PublishResult, error) {
+	m := serialization.Message{}
+	if lc.acceptableContentType != contentType {
+		return PublishResult{}, fmt.Errorf("contentType %q not compatible with expected contentType %q", contentType, lc.acceptableContentType)
```

Comment:
nit: the error message is probably what we intend to support at some point, but it would be quite confusing at the moment if one sets `text/*` on the stream client, receives a message with `text/plain` and reads that error msg

Reply:
Yes, will add a TODO